src/silo/query_engine/operators/selection.cpp

```diff
@@ -0,0 +1,103 @@
+#include "silo/query_engine/operators/selection.h"
+
+#include <roaring/roaring.hh>
+#include <vector>
+
+#include "silo/query_engine/operators/operator.h"
+
+namespace silo::query_engine::operators {
+
+Selection::Selection(
+   uint64_t const* column,
+   Comparator comparator,
+   uint64_t value,
+   unsigned sequence_count
+)
+    : column(column),
+      comparator(comparator),
+      value(value),
+      sequence_count(sequence_count) {}
+
+Selection::~Selection() noexcept = default;
+
+std::string Selection::toString(const Database& /*database*/) const {
+   return "Select";
+}
+
+Type Selection::type() const {
+   return SELECTION;
+}
+
+void Selection::negate() {
+   switch (this->comparator) {
+      case EQUALS:
+         this->comparator = NOT_EQUALS;
+         break;
+      case NOT_EQUALS:
+         this->comparator = EQUALS;
+         break;
+      case LESS:
+         this->comparator = HIGHER_OR_EQUALS;
+         break;
+      case HIGHER_OR_EQUALS:
+         this->comparator = LESS;
+         break;
+      case HIGHER:
+         this->comparator = LESS_OR_EQUALS;
+         break;
+      case LESS_OR_EQUALS:
+         this->comparator = HIGHER;
+         break;
+   }
+}
+
+OperatorResult Selection::evaluate() const {
+   OperatorResult res = {new roaring::Roaring(), nullptr};
+   switch (this->comparator) {
+      case EQUALS:
+         for (unsigned i = 0; i < sequence_count; i++) {
+            if (column[i] == value) {
+               res.mutable_res->add(i);
+            }
+         }
+         break;
+      case NOT_EQUALS:
+         for (unsigned i = 0; i < sequence_count; i++) {
+            if (column[i] != value) {
+               res.mutable_res->add(i);
+            }
+         }
+         break;
+      case LESS:
+         for (unsigned i = 0; i < sequence_count; i++) {
+            if (column[i] < value) {
+               res.mutable_res->add(i);
+            }
+         }
+         break;
+      case HIGHER_OR_EQUALS:
+         for (unsigned i = 0; i < sequence_count; i++) {
+            if (column[i] >= value) {
+               res.mutable_res->add(i);
+            }
+         }
+         break;
+      case HIGHER:
+         for (unsigned i = 0; i < sequence_count; i++) {
+            if (column[i] > value) {
+               res.mutable_res->add(i);
+            }
+         }
+         break;
+      case LESS_OR_EQUALS:
+         for (unsigned i = 0; i < sequence_count; i++) {
+            if (column[i] <= value) {
+               res.mutable_res->add(i);
+            }
+         }
```

Comment:
As the predicate is quite complicated it will not be inlined by the compiler sufficiently.
This loop is actually one that actually iterates over all sequences, so it should be as tight as possible.

For loops that only iterate over operators this is not crucial